// forthright48/world/index.js
(function(){
    "use strict";

    var path = require ( "path" );

    var root = path.join ( __dirname, "../../.." );
    var secret = require ( path.join ( root, "/secret.js" ) );
    var bcryptWorkLoad = 10;

    module.exports = {
        root: root,
        handleError: handleError,
        secret: secret,
        getLayoutContext: getLayoutContext,
        myRender: myRender,
        bcryptWorkLoad: bcryptWorkLoad

    };

    /*******************************************
    Implementation
    *******************************************/

    // Create an object wrapper for errors
    function handleError ( req, res, error, realError ) {
        var context = {
            error: error,
            realError: realError
        };
        var subtitle = "error";
        context.layoutContext = getLayoutContext ( req, subtitle );

        return res.render ( "error", context);
    }

    // Returns Context of Layout
    function getLayoutContext( req, subtitle ) {
        var session = req.session || {};

        var superUser = false;
        if ( req && req.session && req.session.status !== 'user' ) superUser = true;

        var adminMode = false;
        if ( req && req.session && req.session.adminMode ) adminMode = true;

        return {
            header: {
                isLoggedIn: session.isLoggedIn,
                username: session.username,
                subtitle: subtitle,
                superUser: superUser,
                adminMode: adminMode
            }
        };
    }

    function getSubtitle ( view ) {
        if ( view === 'success' ) return view;
        else if ( view === 'home' ) return view;
        else if ( view === 'problem-creation/problem-creation' ) return 'problem-creation';
        else if ( view === 'gateway/gateway') return 'gateway';
        return "";
    }

    // Automatically injects the context of layout
    function myRender ( req, res, view, context ) {
        context = context || {}; ///In case it is empty
        var subtitle = getSubtitle ( view );
        context.layoutContext = getLayoutContext ( req, subtitle );
        context.superUser = context.layoutContext.header.superUser;
        context.adminMode = context.layoutContext.header.adminMode;

        return res.render ( view, context );
    }

}());
